gnd/src/main/java/com/google/android/gnd/system/CameraManager.java

```diff
@@ -50,19 +51,17 @@ public Completable launchImageCapture() {
         .andThen(sendCaptureImageIntent());
   }
 
-  public Observable<Optional<Bitmap>> captureImageResult() {
+  public Observable<Bitmap> captureImageResult() {
     return activityStreams
         .getNextActivityResult(CAPTURE_PHOTO_REQUEST_CODE)
         .filter(ActivityResult::isOk)
         .map(
-            activityResult -> {
-              Intent intent = activityResult.getData();
-              if (intent != null && intent.getExtras() != null) {
-                Bitmap photo = (Bitmap) intent.getExtras().get("data");
-                return Optional.ofNullable(photo);
-              }
-              return Optional.empty();
-            });
+            activityResult ->
+                Optional.ofNullable(activityResult.getData())
+                    .map(Intent::getExtras)
+                    .map(extras -> (Bitmap) extras.get("data")))
+        .filter(Optional::isPresent)
+        .map(Optional::get);
```

Comment:
``` 
  .getNextActivityResult(CAPTURE_PHOTO_REQUEST_CODE)
  .flatMap(this::onCaptureImageResult)
```

Then:
```
private static Observable<Bitmap> onCaptureImageResult(ActivityResult result) {
  return Observable.create(em -> {
    if (!result.isOk()) {
      return;
    }
    Intent data = activityResult.getData();
    if (data == null) {
      return;
    }
    // and so on..
    em.onNext((Bitmap) extras.get("data"));
  });
}
```

Reply:
This would convert our RxChains into flat code. What's the benefit of that?

Reply:
In this case, do you mean it would replace the functional reactive programming (FRP) paradigm with an imperative one?

Functional reactive constructs are very useful for representing flows of data while managing/limiting side effects in an expressive, consistent way. But since Java is primarily imperative, we'll always be jumping back and forth between functional and imperative paradigms anyway. Where we choose to do that is up to us, and imo, one clear place where it makes sense to switch to the imperative paradigm is synchronous code that would be much less readable if declared using the FRP model. At least this is what I've observed working with RxJava over the past two years or so.

@scolsen, who has done more research on this than I have, what are your thoughts?

Reply:
I agree with your points. Fixed